cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,333 @@
+            raise TypeError('Only integer repetitions are allowed.')
+        if protocols.is_measurement(self.circuit):
+            raise NotImplementedError('Loops over measurements are not supported.')
+        new_op = self.copy()
+        new_op._repetitions *= repetitions
+        if modulus:
+            if validate_modulus and not linalg.allclose_up_to_global_phase(
+                protocols.unitary(self.circuit * modulus), np.eye(np.prod(self.circuit.qid_shape()))
+            ):
+                raise ValueError('Raising the circuit to "modulus" ' 'must produce the identity.')
+            # Map repetitions to (-exp_modulus / 2, exp_modulus / 2]
+            new_op._repetitions %= modulus
+            if allow_invert and new_op._repetitions > modulus // 2:
+                new_op._repetitions -= modulus
+        return new_op
+
+    def __pow__(self, power: int) -> 'CircuitOperation':
+        return self.repeat(power)
+
+    def with_qubit_mapping(self, qubit_map: Dict['cirq.Qid', 'cirq.Qid']) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated qubit mapping.
+
+        The provided key_map is composed with the existing map, so calling this
+        function with qubit_map={GridQubit(0, 1): GridQubit(0, 2)} on a
+        CircuitOperation with qubit map {GridQubit(1, 1): GridQubit(0, 1)} will
+        result in a final map of {GridQubit(1, 1): GridQubit(0, 2)}.
+        """
+        new_op = self.copy()
+        new_op._qubit_map = {k: qubit_map.get(v, v) for k, v in new_op.qubit_map.items()}
+        return new_op
+
+    def with_qubits(self, *new_qubits: 'cirq.Qid'):
+        if len(new_qubits) != protocols.num_qubits(self.circuit):
+            expected = protocols.num_qubits(self.circuit)
+            raise ValueError(f'Expected {expected} qubits, got {len(new_qubits)}.')
+        return self.with_qubit_mapping(dict(zip(self.qubits, new_qubits)))
+
+    def with_measurement_key_mapping(self, key_map: Dict[str, str]) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated key mapping.
+
+        The provided key_map is composed with the existing map, so calling this
+        function with key_map={'b': 'c'} on a CircuitOperation with key map
+        {'a': 'b'} will result in a final map of {'a': 'c'}.
+        """
+        new_op = self.copy()
+        new_op._measurement_key_map = {
+            k: key_map.get(v, v) for k, v in new_op._measurement_key_map.items()
+        }
+        return new_op
+
+    def _with_measurement_key_mapping_(self, key_map: Dict[str, str]) -> 'CircuitOperation':
+        return self.with_measurement_key_mapping(key_map)
+
+    def with_params(
+        self, param_values: study.ParamResolverOrSimilarType, recursive: bool = False
+    ) -> 'CircuitOperation':
+        """Returns a copy of this operation with updated param values.
+
+        The provided param_values are composed with the existing values, so
```

Comment:
nit: format docsstring for public method: Args are missing, Return is missing

Reply:
Done.